FLT/Deformations/RepresentationTheory/Basic.lean

```diff
@@ -0,0 +1,379 @@
+
+@[simp]
+lemma GaloisRep.map_conj (ρ : GaloisRep K A M) (e : M ≃ₗ[A] N) (f : K →+* L) :
+    (ρ.conj e).map f = (ρ.map f).conj e := rfl
+
+omit [NumberField K] in
+@[simp]
+lemma GaloisRep.ker_conj (ρ : GaloisRep K A M) (e : M ≃ₗ[A] N) :
+    (ρ.conj e).ker = ρ.ker := by
+  letI := moduleTopology A (Module.End A M)
+  letI := moduleTopology A (Module.End A N)
+  ext; simp [conj]
+
+/-- Equivalent modules have equivalent set of galois reps. -/
+noncomputable
+def GaloisRep.conjEquiv (e : M ≃ₗ[A] N) : GaloisRep K A M ≃ GaloisRep K A N where
+  toFun := (conj · e)
+  invFun := (conj · e.symm)
+  left_inv _ := by ext; simp
+  right_inv _ := by ext; simp
+
+/-- Given a basis, we may frame a galois rep into a framed galois rep. -/
+noncomputable
+def GaloisRep.frame (ρ : GaloisRep K A M) (b : Basis n A M) : FramedGaloisRep K A n :=
+  ρ.conj (b.repr ≪≫ₗ Finsupp.linearEquivFunOnFinite A A n)
+
+/-- Given a basis of `M`, we may realize a framed galois rep as a galois rep on `M`. -/
+noncomputable
+def FramedGaloisRep.unframe (ρ : FramedGaloisRep K A n) (b : Basis n A M) : GaloisRep K A M :=
+  ρ.conj (b.repr ≪≫ₗ Finsupp.linearEquivFunOnFinite A A n).symm
+
+omit [DecidableEq n] [NumberField K] in
+@[simp]
+lemma GaloisRep.unframe_frame (ρ : GaloisRep K A M) (b : Basis n A M) :
+    (ρ.frame b).unframe b = ρ := by
+  ext; simp [frame, FramedGaloisRep.unframe]
+
+omit [DecidableEq n] [NumberField K] in
+@[simp]
+lemma FramedGaloisRep.unframe_frame (ρ : FramedGaloisRep K A n) (b : Basis n A M) :
+    (ρ.unframe b).frame b = ρ := by
+  ext; simp [unframe, GaloisRep.frame]; rfl
+
+variable [IsTopologicalRing A]
+
+/-- `A`-linear framed galois reps are equivalent to continuous homomorphisms into `GLₙ(A)`. -/
+def FramedGaloisRep.GL : FramedGaloisRep K A n ≃ (Γ K →ₜ* GL n A) :=
+  letI := moduleTopology A (Module.End A (n → A))
+  letI : ContinuousMul _ := ⟨IsModuleTopology.continuous_mul_of_finite A (Module.End A (n → A))⟩
+  letI e : Module.End A (n → A) ≃A[A] Matrix n n A :=
+    .ofIsModuleTopology LinearMap.toMatrixAlgEquiv'
+  { toFun ρ := (e.toContinuousAlgHom.toContinuousMonoidHom.comp ρ).toHomUnits
+    invFun ρ := e.symm.toContinuousAlgHom.toContinuousMonoidHom.comp ((Units.coeHomₜ _).comp ρ)
+    left_inv _ := by ext; simp [GaloisRep]
+    right_inv _ := by ext; simp }
+
+omit [NumberField K] in
+@[simp]
+lemma FramedGaloisRep.GL_apply (ρ : FramedGaloisRep K A n) (σ) : (ρ.GL σ).1 = (ρ σ).toMatrix' := rfl
+
+/-- Make an `A`-linear framed galois reps from a continuous hom into `GLₙ(A)`. -/
+abbrev FramedGaloisRep.ofGL := FramedGaloisRep.GL (K := K) (A := A) (n := n).symm
+
+omit [NumberField K] in
+@[simp]
+lemma FramedGaloisRep.GL_symm_apply (ρ : Γ K →ₜ* GL n A) (σ) : GL.symm ρ σ = (ρ σ).toLin := rfl
+
+omit [NumberField K] in
+@[simp]
+lemma FramedGaloisRep.ofGL_apply (ρ : Γ K →ₜ* GL n A) (σ) : ofGL ρ σ = (ρ σ).toLin := rfl
+
+/-- `1`-dimensional framed galois reps are equivalent to (continuous) characters.  -/
```

Comment:
I have some linter complaining about this line (apparently "doc-strings should end with a single space or newline")